// sails-service-mailer/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

exports.default = function (type, config) {
  if (mailer[type.toLowerCase()] instanceof Function) {
    return new mailer[type.toLowerCase()](config);
  } else {
    throw new Error('Unrecognized type -> ' + type);
  }
};

var _DirectMailer = require('./DirectMailer');

var _DirectMailer2 = _interopRequireDefault(_DirectMailer);

var _SendGridMailer = require('./SendGridMailer');

var _SendGridMailer2 = _interopRequireDefault(_SendGridMailer);

var _SendMailMailer = require('./SendMailMailer');

var _SendMailMailer2 = _interopRequireDefault(_SendMailMailer);

var _SESMailer = require('./SESMailer');

var _SESMailer2 = _interopRequireDefault(_SESMailer);

var _SMTPMailer = require('./SMTPMailer');

var _SMTPMailer2 = _interopRequireDefault(_SMTPMailer);

var _StubMailer = require('./StubMailer');

var _StubMailer2 = _interopRequireDefault(_StubMailer);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var mailer = {
  direct: _DirectMailer2.default,
  sendgrid: _SendGridMailer2.default,
  sendmail: _SendMailMailer2.default,
  ses: _SESMailer2.default,
  smtp: _SMTPMailer2.default,
  stub: _StubMailer2.default
};

/**
 * Create mailer instance based on type
 * @param {String} type Type of mailer
 * @param {Object} [config] Configuration object
 */
module.exports = exports['default'];
